test/Interpreter/bridged_casts_folding.swift

```diff
@@ -0,0 +1,641 @@
+  // CHECK: stderr>>> Could not cast value of type '__NSCFNumber' (0x{{[0-9a-f]*}}) to 'NSString' (0x{{[0-9a-f]*}}).
+  // CHECK: stderr>>> OK: saw expected "crashed: sigabrt"
+  // CHECK: [       OK ] BridgedCastFolding.NSArray (Int) -> Swift Array (String). Crashing.
+
+  // CHECK-OPT-LABEL: [ RUN      ] BridgedCastFolding.NSArray (Int) -> Swift Array (String). Crashing.
+  // CHECK-OPT: stderr>>> Could not cast value of type '__NSCFNumber' (0x{{[0-9a-f]*}}) to 'NSString' (0x{{[0-9a-f]*}}).
+  // CHECK-OPT: stderr>>> OK: saw expected "crashed: sigabrt"
+  // CHECK-OPT: [       OK ] BridgedCastFolding.NSArray (Int) -> Swift Array (String). Crashing.
+  expectCrashLater()
+  do {
+    let arr: [String] = forcedCast(nsArrInt)
+    expectEqual(arr, swiftArrString)
+  }
+}
+
+// Check optimization of casts from NSDictionary to Swift Dictionary
+
+let swiftDictInt: [Int: Int] = [1:1, 2:2, 3:3, 4:4]
+let swiftDictDouble: [Double: Double] = [1.1 : 1.1, 2.2 : 2.2, 3.3 : 3.3, 4.4 : 4.4]
+let swiftDictString: [String: String] = ["One":"One", "Two":"Two", "Three":"Three", "Four":"Four"]
+let nsDictInt: NSDictionary = [1:1, 2:2, 3:3, 4:4]
+let nsDictDouble: NSDictionary = [1.1 : 1.1, 2.2 : 2.2, 3.3 : 3.3, 4.4 : 4.4]
+let nsDictString: NSDictionary = ["One":"One", "Two":"Two", "Three":"Three", "Four":"Four"]
+let cfDictInt: CFDictionary = [1:1, 2:2, 3:3, 4:4] as CFDictionary
+let cfDictDouble: CFDictionary = [1.1 : 1.1, 2.2 : 2.2, 3.3 : 3.3, 4.4 : 4.4] as CFDictionary
+let cfDictString: CFDictionary = ["One":"One", "Two":"Two", "Three":"Three", "Four":"Four"] as CFDictionary
+
+Tests.test("NSDictionary -> Swift (Dictionary)") {
+  do {
+    let dict: [Int: Int] = forcedCast(nsDictInt)
+    expectEqual(dict, swiftDictInt)
+  }
+
+  do {
+    let dictOpt: [Int: Int]? = condCast(nsDictInt)
+    expectEqual(dictOpt!, swiftDictInt)
+  }
+
+  do {
+    let dict: [Double: Double] = forcedCast(nsDictDouble)
+    expectEqual(dict, swiftDictDouble)
+  }
+
+  do {
+    let dictOpt: [Double: Double]? = condCast(nsDictDouble)
+    expectEqual(dictOpt!, swiftDictDouble)
+  }
+
+  do {
+    let dict: [String: String] = forcedCast(nsDictString)
+    expectEqual(dict, swiftDictString)
+  }
+
+  do {
+    let dictOpt: [String: String]? = condCast(nsDictString)
+    expectEqual(dictOpt!, swiftDictString)
+  }
+
+  do {
+    let dictOpt: [Int: Int]? = condCast(nsDictString)
+    expectNil(dictOpt)
+  }
+}
+
+Tests.test("NSDictionary -> Swift (Dictionary). Crashing Test Cases") {
+  do {
+    // Will this crash?
+    let dictOpt: [Int: Int]? = condCast(nsDictString)
+    expectNil(dictOpt)
+  }
+
+  // NOTE: This is preserving current behavior. We should probably have a better
+  // error here like in the other crashes.
+  //
+  // CHECK-LABEL: [ RUN      ] BridgedCastFolding.NSDictionary -> Swift (Dictionary). Crashing Test Cases
+  // CHECK: stderr>>> OK: saw expected "crashed: sigill"
```

Comment:
@ravikandhadai This is a cast where I think we could give a better error message. (Really we should see if we can de-obfuscate these types at runtime. If we know we are going to crash, we should see if we can do a better job with this since runtime isn't an issue. Might be interesting for you guys).

Reply:
Sorry, let me be more clear. I was commenting on two things here:

1. I think we are optimizing this at -Onone when we should not be, not because we can't, but because we want to leave in the error messages. Today the cast optimizer always tries to optimize these at -Onone via constant propagation. This I think violates the stronger notion of as-if at -Onone. It might be interesting for you to look at.
2. In general the runtime error messages that I found here have internal runtime names. I wonder if it would make sense to use for instance __NSCFString to just say NSString.

Reply:
@ravikandhadai  (I might fix #1 if I have time).

Reply:
@gottesmm I am not very familiar with this  aspect of constant propagation. Just to get more clarity on the problem, is`forcedCast(nsDictString)` being optimized by ConstantPropagation? It is not transparent, so I am bit confused here.

Reply:
The missing message turned out to be caused by an issue in the Foundation overlay (fixed in https://github.com/apple/swift/pull/23174).

Reply:
Thanks for the update.